GPU/Vulkan/FramebufferVulkan.cpp

```diff
@@ -812,30 +894,98 @@ struct CardboardSettings * FramebufferManagerVulkan::GetCardboardSettings(struct
 	return nullptr;
 }
 
-void FramebufferManagerVulkan::CopyDisplayToOutput() {
-	// This is where we should collect all the renderpasses from this frame,
-	// sort them in order according to texturing dependencies, and enqueue
-	// them on the Vulkan context.
+void FramebufferManagerVulkan::RecurseRenderPasses(FBRenderPass *rp, std::vector<FBRenderPass *> &order) {
+	if (!rp) {
+		return;
+	}
+	for (auto dep : rp->dependencies) {
+		RecurseRenderPasses(dep, order);
+	}
+	rp->executed = true;
+	order.push_back(rp);
+}
+
+FBRenderPass *FramebufferManagerVulkan::GetRenderPass(VirtualFramebuffer *vfb) {
+	for (auto rp : passes_) {
+		if (rp->vfb == vfb) {
+			return rp;
+		}
+	}
+	return nullptr;
+}
+
+void FramebufferManagerVulkan::WalkRenderPasses(VirtualFramebuffer *toDisplay, std::vector<FBRenderPass *> &rpOrder) {
+	rpOrder.clear();
+	// The found render passes are marked with executed = true.
+	FBRenderPass *currentRP = GetRenderPass(toDisplay);
+	if (currentRP) {
+		RecurseRenderPasses(currentRP, rpOrder);
+	}
+
+	// Now, see if there are any "orphan" render passes left to cap off and enqueue.
+	for (auto rp : passes_) {
+		if (!rp->executed) {
+			rpOrder.push_back(rp);
+			rp->executed = true;
+		}
+	}
 	
+	// OK, now we should have the complete render pass order.
+}
+
+void FramebufferManagerVulkan::SubmitRenderPasses(std::vector<FBRenderPass *> &passOrder, VirtualFramebuffer *toDisplay) {
+	// TODO: Maybe there's some more optimal way.
+	FBRenderPass *last = nullptr;
+	for (auto &rp : passOrder) {
+		if (rp == last) {
+			DebugBreak();
+		}
+		last = rp;
+		switch (rp->state) {
+		case RPState::RECORDING:  // Should usually only have one left in this state - the main RT from the last frame.
+		{
+			VkCommandBuffer cmd = rp->cmd;
+			vkCmdEndRenderPass(cmd);
+			if (rp->vfb == toDisplay) {
+				rp->vfb->fbo_vk->GetColor()->Transition(cmd, VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL);
+			}
+			vkEndCommandBuffer(cmd);
+			rp->state = RPState::SUBMITTED;
+			// TODO: Collect transitions from dependencies of the next renderpass and 
+			// tack them onto rp->cmd here before queueing it.
+			vulkan_->QueueBeforeSurfaceRender(rp->cmd);
+			break;
+		}
+		case RPState::CAPPED:
+			DebugBreak();
+			break;
+		case RPState::PAUSED:
+			// Should we keep it around for the next frame?
```

Comment:
Hmm.  Do we not need to `vkCmdEndRenderPass` it and etc.?  We may try to delete things (soonish) if we keep it around too long without drawing it.

-[Unknown]